File: CS/DXServiceHelper/DXServiceReferenceExtensions.cs
namespace DXServiceHelper.DXService {
    public partial class ViewProperty {
        public ViewProperty (string fName, CriteriaOperator fProperty, SortDirection fSorting, bool fGroup,
            bool fFetch) {
            Name = fName;
            Property = fProperty;
            Sorting = fSorting;
            Group = fGroup;
            Fetch = fFetch;
        }

        public ViewProperty (string fName, CriteriaOperator fProperty, SortDirection fSorting, bool fGroup)
            : this(fName, fProperty, fSorting, fGroup, true) { }

        public ViewProperty (string fName, CriteriaOperator fProperty, SortDirection fSorting)
            : this(fName, fProperty, fSorting, false) { }

        public ViewProperty (string fName, CriteriaOperator fProperty) : this(fName, fProperty, SortDirection.None) { }
    }

    public partial class AggregateOperand {
        public AggregateOperand (OperandProperty fCollectionProperty, CriteriaOperator fCondition,
            CriteriaOperator fAggregatedExpression, Aggregate fAggregateType) {
            CollectionProperty = fCollectionProperty;
            Condition = fCondition;
            AggregatedExpression = fAggregatedExpression;
            AggregateType = fAggregateType;
        }

        public AggregateOperand (OperandProperty fCollectionProperty, CriteriaOperator fCondition,
            CriteriaOperator fAggregatedExpression)
            : this(fCollectionProperty, fCondition, fAggregatedExpression, Aggregate.Sum) { }

        public AggregateOperand () : base() { }
    }

    public partial class BetweenOperator {
        public BetweenOperator (CriteriaOperator fBeginExpression, CriteriaOperator fEndExpression,
            CriteriaOperator fTestExpression) {
            BeginExpression = fBeginExpression;
            EndExpression = fEndExpression;
            TestExpression = fTestExpression;
        }

        public BetweenOperator () : base() { }
    }

    public partial class BinaryOperator {
        public BinaryOperator (CriteriaOperator fLeftOperand, CriteriaOperator fRightOperand,
            BinaryOperatorType fOperatorType) {
            LeftOperand = fLeftOperand;
            RightOperand = fRightOperand;
            OperatorType = fOperatorType;
        }

        public BinaryOperator (CriteriaOperator fLeftOperand, CriteriaOperator fRightOperand)
            : this(fLeftOperand, fRightOperand, BinaryOperatorType.Equal) { }

        public BinaryOperator () : base() { }
    }

    public partial class FunctionOperator {
        public FunctionOperator (CriteriaOperator[] fOperands, FunctionOperatorType fOperatorType) {
            Operands = fOperands;
            OperatorType = fOperatorType;
        }

        public FunctionOperator () : base() { }
    }

    public partial class GroupOperator {
        public GroupOperator (CriteriaOperator[] fOperands, GroupOperatorType fOperatorType) {
            Operands = fOperands;
            OperatorType = fOperatorType;
        }

        public GroupOperator (CriteriaOperator[] fOperands) : this(fOperands, GroupOperatorType.And) { }

        public GroupOperator () : base() { }
    }

    public partial class InOperator {
        public InOperator (CriteriaOperator fLeftOperand, CriteriaOperator[] fOperands) {
            LeftOperand = fLeftOperand;
            Operands = fOperands;
        }

        public InOperator () : base() { }
    }

    public partial class JoinNode {
        public JoinNode (JoinNode[] fSubNodes, CriteriaOperator fCondition, string fAlias, JoinType fType,
            string fTableName) {
            SubNodes = fSubNodes;
            Condition = fCondition;
            Alias = fAlias;
            Type = fType;
            TableName = fTableName;
        }

        public JoinNode (JoinNode[] fSubNodes, CriteriaOperator fCondition, string fAlias, string fTableName)
            : this(fSubNodes, fCondition, fAlias, JoinType.Inner, fTableName) { }
    }

    public partial class JoinOperand {
        public JoinOperand (CriteriaOperator fCondition, string fJoinTypeName, CriteriaOperator fAggregatedExpression,
            Aggregate fAggregateType) {
            Condition = fCondition;
            JoinTypeName = fJoinTypeName;
            AggregatedExpression = fAggregatedExpression;
            AggregateType = fAggregateType;
        }

        public JoinOperand (CriteriaOperator fCondition, string fJoinTypeName,
            CriteriaOperator fAggregatedExpression)
            : this(fCondition, fJoinTypeName, fAggregatedExpression, Aggregate.Sum) { }

        public JoinOperand () : base() { }
    }

    public partial class OperandProperty {
        public OperandProperty (string fPropertyName) { PropertyName = fPropertyName; }

        public OperandProperty () : base() { }
    }

    public partial class OperandValue {
        public OperandValue (object fItem) { Item = fItem; }

        public OperandValue () : base() { }
    }

    public partial class ParameterValue {
        public ParameterValue (object fItem, int fTag) : base(fItem) { Tag = fTag; }

        public ParameterValue () : base() { }
    }

    public partial class QueryOperand {
        public QueryOperand (string fColumnName, DBColumnType fColumnType, string fNodeAlias) {
            ColumnName = fColumnName;
            ColumnType = fColumnType;
            NodeAlias = fNodeAlias;
        }

        public QueryOperand (string fColumnName, string fNodeAlias)
            : this(fColumnName, DBColumnType.String, fNodeAlias) { }

        public QueryOperand () : base() { }
    }

    public partial class QuerySubQueryContainer {
        public QuerySubQueryContainer (BaseStatement fNode, CriteriaOperator fAggregateProperty,
            Aggregate fAggregateType) {
            Node = fNode;
            AggregateProperty = fAggregateProperty;
            AggregateType = fAggregateType;
        }

        public QuerySubQueryContainer (BaseStatement fNode, CriteriaOperator fAggregateProperty)
            : this(fNode, fAggregateProperty, Aggregate.Sum) { }

        public QuerySubQueryContainer () : base() { }
    }

    public partial class UnaryOperator {
        public UnaryOperator (CriteriaOperator fOperand, UnaryOperatorType fOperatorType) {
            Operand = fOperand;
            OperatorType = fOperatorType;
        }

        public UnaryOperator () : base() { }
    }
}

File: VB/DXServiceHelper/DXServiceReferenceExtensions.vb
Imports Microsoft.VisualBasic
Imports System
Namespace DXService
    Partial Public Class ViewProperty
        Public Sub New(ByVal fName As String, ByVal fProperty As CriteriaOperator, ByVal fSorting As SortDirection, ByVal fGroup As Boolean, ByVal fFetch As Boolean)
            Name = fName
            [Property] = fProperty
            Sorting = fSorting
            Group = fGroup
            Fetch = fFetch
        End Sub

        Public Sub New(ByVal fName As String, ByVal fProperty As CriteriaOperator, ByVal fSorting As SortDirection, ByVal fGroup As Boolean)
            Me.New(fName, fProperty, fSorting, fGroup, True)
        End Sub

        Public Sub New(ByVal fName As String, ByVal fProperty As CriteriaOperator, ByVal fSorting As SortDirection)
            Me.New(fName, fProperty, fSorting, False)
        End Sub

        Public Sub New(ByVal fName As String, ByVal fProperty As CriteriaOperator)
            Me.New(fName, fProperty, SortDirection.None)
        End Sub
    End Class

    Partial Public Class AggregateOperand
        Public Sub New(ByVal fCollectionProperty As OperandProperty, ByVal fCondition As CriteriaOperator, ByVal fAggregatedExpression As CriteriaOperator, ByVal fAggregateType As Aggregate)
            CollectionProperty = fCollectionProperty
            Condition = fCondition
            AggregatedExpression = fAggregatedExpression
            AggregateType = fAggregateType
        End Sub

        Public Sub New(ByVal fCollectionProperty As OperandProperty, ByVal fCondition As CriteriaOperator, ByVal fAggregatedExpression As CriteriaOperator)
            Me.New(fCollectionProperty, fCondition, fAggregatedExpression, Aggregate.Sum)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class BetweenOperator
        Public Sub New(ByVal fBeginExpression As CriteriaOperator, ByVal fEndExpression As CriteriaOperator, ByVal fTestExpression As CriteriaOperator)
            BeginExpression = fBeginExpression
            EndExpression = fEndExpression
            TestExpression = fTestExpression
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class BinaryOperator
        Public Sub New(ByVal fLeftOperand As CriteriaOperator, ByVal fRightOperand As CriteriaOperator, ByVal fOperatorType As BinaryOperatorType)
            LeftOperand = fLeftOperand
            RightOperand = fRightOperand
            OperatorType = fOperatorType
        End Sub

        Public Sub New(ByVal fLeftOperand As CriteriaOperator, ByVal fRightOperand As CriteriaOperator)
            Me.New(fLeftOperand, fRightOperand, BinaryOperatorType.Equal)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class FunctionOperator
        Public Sub New(ByVal fOperands() As CriteriaOperator, ByVal fOperatorType As FunctionOperatorType)
            Operands = fOperands
            OperatorType = fOperatorType
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class GroupOperator
        Public Sub New(ByVal fOperands() As CriteriaOperator, ByVal fOperatorType As GroupOperatorType)
            Operands = fOperands
            OperatorType = fOperatorType
        End Sub

        Public Sub New(ByVal fOperands() As CriteriaOperator)
            Me.New(fOperands, GroupOperatorType.And)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class InOperator
        Public Sub New(ByVal fLeftOperand As CriteriaOperator, ByVal fOperands() As CriteriaOperator)
            LeftOperand = fLeftOperand
            Operands = fOperands
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class JoinNode
        Public Sub New(ByVal fSubNodes() As JoinNode, ByVal fCondition As CriteriaOperator, ByVal fAlias As String, ByVal fType As JoinType, ByVal fTableName As String)
            SubNodes = fSubNodes
            Condition = fCondition
            [Alias] = fAlias
            Type = fType
            TableName = fTableName
        End Sub

        Public Sub New(ByVal fSubNodes() As JoinNode, ByVal fCondition As CriteriaOperator, ByVal fAlias As String, ByVal fTableName As String)
            Me.New(fSubNodes, fCondition, fAlias, JoinType.Inner, fTableName)
        End Sub
    End Class

    Partial Public Class OperandProperty
        Public Sub New(ByVal fPropertyName As String)
            PropertyName = fPropertyName
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class OperandValue
        Public Sub New(ByVal fItem As Object)
            Item = fItem
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class ParameterValue
        Public Sub New(ByVal fItem As Object, ByVal fTag As Integer)
            MyBase.New(fItem)
            Tag = fTag
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class QueryOperand
        Public Sub New(ByVal fColumnName As String, ByVal fColumnType As DBColumnType, ByVal fNodeAlias As String)
            ColumnName = fColumnName
            ColumnType = fColumnType
            NodeAlias = fNodeAlias
        End Sub

        Public Sub New(ByVal fColumnName As String, ByVal fNodeAlias As String)
            Me.New(fColumnName, DBColumnType.String, fNodeAlias)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class QuerySubQueryContainer
        Public Sub New(ByVal fNode As BaseStatement, ByVal fAggregateProperty As CriteriaOperator, ByVal fAggregateType As Aggregate)
            Node = fNode
            AggregateProperty = fAggregateProperty
            AggregateType = fAggregateType
        End Sub

        Public Sub New(ByVal fNode As BaseStatement, ByVal fAggregateProperty As CriteriaOperator)
            Me.New(fNode, fAggregateProperty, Aggregate.Sum)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class

    Partial Public Class UnaryOperator
        Public Sub New(ByVal fOperand As CriteriaOperator, ByVal fOperatorType As UnaryOperatorType)
            Operand = fOperand
            OperatorType = fOperatorType
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub
    End Class
End Namespace